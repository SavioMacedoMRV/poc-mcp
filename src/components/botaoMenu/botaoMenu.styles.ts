import {ReactNode} from 'react'
import styled from 'styled-components'
import {cores} from '../../resources/cores'
import {margens} from 'resources/margens'
import {hexToRgba} from 'helpers/cores'
import {BotaoIcone} from 'components/botaoIcone'
import {Texto} from 'components/texto'
import {Icone} from 'components/icone'
import {TextoProps} from 'components/texto/texto.styles'
import ChevronSVG from 'icons/iChevron.svg'

export interface BotaoMenuProps {
  desabilitado?: boolean
  icone?: React.FC<React.SVGProps<SVGSVGElement>>
  botao?: {texto: string; textoProps?: TextoProps}
  componenteOpcoes: ReactNode
  cor?: string
  titulo?: string
  largura?: number
  corAtivo?: string
  'data-testid'?: string
}

export const BotaoMenuIcone = styled(BotaoIcone).attrs(props => ({
  altura: 24,
  cor: cores.neutralDark,
  ...props,
}))``

export const BotaoMenuTexto = styled.button<{
  $corBorda?: string
}>`
  display: flex;
  flex: 1;
  width: 160px;
  border-radius: ${margens.xsmall}px;
  border: 2px solid ${({$corBorda}) => $corBorda ?? cores.background03};
  background: none;
  align-items: center;
  gap: ${margens.xsmall}px;
  flex-direction: row;
  justify-content: flex-start;
  padding: ${margens.xxsmall}px ${margens.xsmall}px;
  cursor: pointer;

  &:hover {
    opacity: 0.7;
  }

  &:active {
    opacity: ${({disabled}) => (disabled ? 0.7 : 1)};
  }

  &:disabled {
    opacity: 0.7;
    cursor: default;
  }

  &:focus-visible {
    outline: 2px solid ${cores.neutralMedium};
    border-radius: ${margens.xsmall}px;
  }
`

export const ContainerBotaoMenu = styled.div`
  display: flex;
  position: relative;
`

export const ContainerSuspenso = styled.div`
  display: flex;
  position: absolute;
  flex-direction: column;
  min-width: 250px;
  max-width: 300px;
  top: 100%;
  right: 50%;
  transform: translateX(50%);
  z-index: 1;
  margin-top: ${margens.xsmall}px;
  background: ${cores.neutralXXLight};
  padding: ${margens.medium}px;
  border-radius: ${margens.small}px;
  box-shadow: 0px 8px 24px 0px ${hexToRgba(cores.neutralXDark, 0.35)};
`

export const IconeBotao = styled(Icone).attrs({
  altura: 24,
  cor: cores.neutralDark,
  icone: ChevronSVG,
})``

export const TextoBotao = styled(Texto).attrs(props => ({
  cor: cores.neutralDark,
  tamanho: 12,
  estilo: 'semibold',
  cursor: 'pointer',
  ...props,
}))`
  text-overflow: ellipsis;
`

export const TextoTitulo = styled(Texto).attrs({
  cor: cores.neutralDark,
  estilo: 'semibold',
  tamanho: 16,
})`
  margin-bottom: ${margens.medium}px;
`
