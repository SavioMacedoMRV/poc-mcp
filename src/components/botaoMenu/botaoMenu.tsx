import React, {useEffect, useRef, useState} from 'react'
import {StyleSheetManager} from 'styled-components'
import {
  BotaoMenuProps,
  BotaoMenuIcone,
  ContainerBotaoMenu,
  ContainerSuspenso,
  BotaoMenuTexto,
  IconeBotao,
  TextoBotao,
  TextoTitulo,
} from './botaoMenu.styles'
import {cores} from 'resources/cores'

export const BotaoMenu = ({
  botao,
  icone,
  titulo,
  cor,
  corAtivo,
  desabilitado,
  componenteOpcoes,
  'data-testid': testid,
  ...props
}: Readonly<BotaoMenuProps>) => {
  const [exibirOpcoes, setExibirOpcoes] = useState(false)
  const dropdownRef = useRef<HTMLDivElement>(null)

  const handleChangeExibirOpcoes = () => {
    setExibirOpcoes(prev => !prev)
  }

  const handleCliqueForaDoMenu = (event: MouseEvent) => {
    if (
      dropdownRef.current &&
      !dropdownRef.current.contains(event.target as Node)
    ) {
      setExibirOpcoes(false)
    }
  }

  useEffect(() => {
    document.addEventListener('mousedown', handleCliqueForaDoMenu)
    return () => {
      document.removeEventListener('mousedown', handleCliqueForaDoMenu)
    }
  }, [])

  let corBotao = cor
  if (exibirOpcoes) {
    corBotao = corAtivo ?? cores.complementary05
  }
  return (
    <StyleSheetManager shouldForwardProp={prop => !!prop}>
      <ContainerBotaoMenu ref={dropdownRef} data-testid={testid} {...props}>
        {icone ? (
          <BotaoMenuIcone
            data-testid={
              testid ? `${testid}-icon-exibir-opcoes` : 'btn-icon-exibir-opcoes'
            }
            aoClicar={handleChangeExibirOpcoes}
            icone={icone}
            cor={corBotao}
            disabled={desabilitado}
          />
        ) : (
          botao && (
            <BotaoMenuTexto
              data-testid={
                testid ? `${testid}-exibir-opcoes` : 'btn-exibir-opcoes'
              }
              onClick={handleChangeExibirOpcoes}
              disabled={desabilitado}
              $corBorda={corBotao}
              {...props}
            >
              <IconeBotao cor={corBotao} inclinacao={exibirOpcoes ? 180 : 0} />
              <TextoBotao
                data-testid={'txt-botao-menu'}
                cursor={desabilitado ? 'default' : 'pointer'}
                {...botao.textoProps}
              >
                {botao.texto}
              </TextoBotao>
            </BotaoMenuTexto>
          )
        )}
        {exibirOpcoes && (
          <ContainerSuspenso
            data-testid="menu-suspenso"
            onClick={handleChangeExibirOpcoes}
          >
            {titulo && <TextoTitulo>{titulo}</TextoTitulo>}
            {componenteOpcoes}
          </ContainerSuspenso>
        )}
      </ContainerBotaoMenu>
    </StyleSheetManager>
  )
}
