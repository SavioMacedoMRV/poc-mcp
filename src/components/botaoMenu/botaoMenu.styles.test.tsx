import {render} from '@testing-library/react'
import * as S from './botaoMenu.styles'

describe('BotaoMenu.styles', () => {
  test('BotaoMenuIcone DEVE ser igual ao snapshot', () => {
    const {container} = render(
      <S.BotaoMenuIcone aoClicar={jest.fn()} icone={() => <svg />} />,
    )
    expect(container).toMatchSnapshot()
  })

  test('BotaoMenuTexto DEVE ser igual ao snapshot', () => {
    const {container} = render(<S.BotaoMenuTexto />)
    expect(container).toMatchSnapshot()
  })

  test('ContainerBotaoMenu DEVE ser igual ao snapshot', () => {
    const {container} = render(<S.ContainerBotaoMenu />)
    expect(container).toMatchSnapshot()
  })

  test('ContainerSuspenso DEVE ser igual ao snapshot', () => {
    const {container} = render(<S.ContainerSuspenso />)
    expect(container).toMatchSnapshot()
  })

  test('IconeBotao DEVE ser igual ao snapshot', () => {
    const {container} = render(<S.IconeBotao />)
    expect(container).toMatchSnapshot()
  })

  test('IconeBotao DEVE ser igual ao snapshot', () => {
    const {container} = render(<S.IconeBotao />)
    expect(container).toMatchSnapshot()
  })

  test('TextoBotao DEVE ser igual ao snapshot', () => {
    const {container} = render(<S.TextoBotao />)
    expect(container).toMatchSnapshot()
  })

  test('TextoTitulo DEVE ser igual ao snapshot', () => {
    const {container} = render(<S.TextoTitulo />)
    expect(container).toMatchSnapshot()
  })
})
