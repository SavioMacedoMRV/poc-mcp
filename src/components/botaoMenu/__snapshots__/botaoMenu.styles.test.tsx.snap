// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BotaoMenu.styles BotaoMenuIcone DEVE ser igual ao snapshot 1`] = `
.c0 {
  display: flex;
  border-radius: 8px;
  border: none;
  align-items: center;
  justify-content: center;
  background: transparent;
  padding: 8px;
  opacity: 1;
  cursor: pointer;
}

.c0:hover {
  opacity: 0.7;
}

.c0:active {
  opacity: 1;
}

.c0:focus-visible {
  outline: unset;
  border: 2px solid #999E9C;
}

<div>
  <button
    class="c0 "
    cor="#5D615F"
  >
    <svg />
  </button>
</div>
`;

exports[`BotaoMenu.styles BotaoMenuTexto DEVE ser igual ao snapshot 1`] = `
.c0 {
  display: flex;
  flex: 1;
  width: 160px;
  border-radius: 8px;
  border: 2px solid #F2F2F2;
  background: none;
  align-items: center;
  gap: 8px;
  flex-direction: row;
  justify-content: flex-start;
  padding: 4px 8px;
  cursor: pointer;
}

.c0:hover {
  opacity: 0.7;
}

.c0:active {
  opacity: 1;
}

.c0:disabled {
  opacity: 0.7;
  cursor: default;
}

.c0:focus-visible {
  outline: 2px solid #999E9C;
  border-radius: 8px;
}

<div>
  <button
    class="c0"
  />
</div>
`;

exports[`BotaoMenu.styles ContainerBotaoMenu DEVE ser igual ao snapshot 1`] = `
.c0 {
  display: flex;
  position: relative;
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`BotaoMenu.styles ContainerSuspenso DEVE ser igual ao snapshot 1`] = `
.c0 {
  display: flex;
  position: absolute;
  flex-direction: column;
  min-width: 250px;
  max-width: 300px;
  top: 100%;
  right: 50%;
  transform: translateX(50%);
  z-index: 1;
  margin-top: 8px;
  background: #FFFFFF;
  padding: 24px;
  border-radius: 16px;
  box-shadow: 0px 8px 24px 0px rgba(67, 70, 69, 0.35);
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`BotaoMenu.styles IconeBotao DEVE ser igual ao snapshot 1`] = `
<div>
  <svg
    class=""
    data-file-name="SvgIChevron"
    fill="#5D615F"
    height="24"
    style="min-width: 24px; min-height: 24px; max-width: 24px; max-height: 24px; transform: rotate(0deg);"
    width="24"
  />
</div>
`;

exports[`BotaoMenu.styles IconeBotao DEVE ser igual ao snapshot 2`] = `
<div>
  <svg
    class=""
    data-file-name="SvgIChevron"
    fill="#5D615F"
    height="24"
    style="min-width: 24px; min-height: 24px; max-width: 24px; max-height: 24px; transform: rotate(0deg);"
    width="24"
  />
</div>
`;

exports[`BotaoMenu.styles TextoBotao DEVE ser igual ao snapshot 1`] = `
.c0 {
  font-weight: 400;
  font-size: 12px;
  line-height: normal;
  font-family: AvertaStd-Semibold;
  color: #5D615F;
  text-align: start;
  cursor: pointer;
  text-transform: none;
}

.c1 {
  text-overflow: ellipsis;
}

<div>
  <span
    class="c0 c1"
    cor="#5D615F"
    cursor="pointer"
    estilo="semibold"
    tamanho="12"
  />
</div>
`;

exports[`BotaoMenu.styles TextoTitulo DEVE ser igual ao snapshot 1`] = `
.c0 {
  font-weight: 400;
  font-size: 16px;
  line-height: normal;
  font-family: AvertaStd-Semibold;
  color: #5D615F;
  text-align: start;
  cursor: text;
  text-transform: none;
}

.c1 {
  margin-bottom: 24px;
}

<div>
  <span
    class="c0 c1"
    cor="#5D615F"
    estilo="semibold"
    tamanho="16"
  />
</div>
`;
